File: testes/algoritmos-ordenacao/merge-sort.js
export function merge(left, right) {
    let vetor = [];

    while (left.length && right.length) {
        if (left[0] < right[0]) {
            vetor.push(left.shift());
        } else {
            vetor.push(right.shift());
        }
    }

    return [...vetor, ...left, ...right];
}

export function mergeSort(vetor) {
    const half = vetor.length / 2;

    if (vetor.length < 2) {
        return vetor;
    }

    const left = vetor.splice(0, half);
    return merge(mergeSort(left), mergeSort(vetor));
}

export function executarMergeSort(numeroRepeticoes, vetor, tipo_instancia) {
    let somaTempos = 0;
    let somaConsumosMemoria = 0;
    let mediaTempos = 0;
    let mediaConsumosMemoria = 0;
    let temposMergeSort = [];
    let consumosMemoriaMergeSort = [];

    for (let i = 0; i < numeroRepeticoes; i++) {
        let tempoInicial = 0;
        let tempoFinal = 0;
        let memoriaInicial = 0;
        let memoriaFinal = 0;

        memoriaInicial = process.memoryUsage().heapUsed;
        tempoInicial = performance.now().toFixed(6);
        mergeSort(vetor);
        tempoFinal = performance.now().toFixed(6);
        memoriaFinal = process.memoryUsage().heapUsed;

        temposMergeSort.push((tempoFinal - tempoInicial));
        consumosMemoriaMergeSort.push((memoriaFinal - memoriaInicial));
    }

    //Cálculo da média de tempo gasto
    for (let i = 0; i < temposMergeSort.length; i++) {
        somaTempos += temposMergeSort[i];
    }
    mediaTempos = (somaTempos / temposMergeSort.length);

    //Cálculo da média de consumo de memória
    for (let i = 0; i < consumosMemoriaMergeSort.length; i++) {
        somaConsumosMemoria += consumosMemoriaMergeSort[i];
    }
    mediaConsumosMemoria = (somaConsumosMemoria / consumosMemoriaMergeSort.length);

    //Log de saída
    console.log(`
        Algoritmo: Merge Sort
        Tipo da Instância/Tamanho: ${tipo_instancia}/${vetor.length}
        Qtd. de Iterações: ${numeroRepeticoes}
        Média de Tempo Gasto: ${mediaTempos.toFixed(6)} ms
        Média de Consumo de Memória: ${mediaConsumosMemoria.toFixed(4)} bytes
    `);
}

File: testes/algoritmos-ordenacao/python/merge-sort.py

def merge(left, right):
    vetor = []

    while left and right:
        if left[0] < right[0]:
            vetor.append(left.pop(0))
        else:
            vetor.append(right.pop(0))

    return vetor + left + right

def merge_sort(vetor):
    if len(vetor) < 2:
        return vetor

    meio = len(vetor) // 2
    left = vetor[:meio]
    right = vetor[meio:]

    return merge(merge_sort(left), merge_sort(right))

def executar_merge_sort(numero_repeticoes, vetor, tipo_instancia):
    soma_tempos = 0
    soma_consumos_memoria = 0
    media_tempos = 0
    media_consumos_memoria = 0
    tempos_merge_sort = []
    consumos_memoria_merge_sort = []

    for i in range(numero_repeticoes):
        tempo_inicial = 0
        tempo_final = 0
        memoria_inicial = 0
        memoria_final = 0

        memoria_inicial = sys.getsizeof(vetor)
        tempo_inicial = time.time()
        merge_sort(vetor.copy())
        tempo_final = time.time()
        memoria_final = sys.getsizeof(vetor)

        tempos_merge_sort.append((tempo_final - tempo_inicial) * 1000)  # Em milissegundos
        consumos_memoria_merge_sort.append(memoria_final - memoria_inicial)

    soma_tempos = sum(tempos_merge_sort)
    media_tempos = soma_tempos / numero_repeticoes

    soma_consumos_memoria = sum(consumos_memoria_merge_sort)
    media_consumos_memoria = soma_consumos_memoria / numero_repeticoes

    # Log de saída
    print(f"""
        Algoritmo: Merge Sort
        Tipo da Instância/Tamanho: {tipo_instancia}/{len(vetor)}
        Qtd. de Iterações: {numero_repeticoes}
        Média de Tempo Gasto: {media_tempos:.6f} ms
        Média de Consumo de Memória: {media_consumos_memoria:.4f} bytes
    """)
